baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/PreferCollectionConstructors.java

```diff
@@ -0,0 +1,348 @@
+                    .put(NEW_COPY_ON_WRITE_ARRAY_LIST_WITH_ITERABLE, CopyOnWriteArrayList.class)
+                    .put(NEW_HASH_MAP_WITH_MAP, HashMap.class)
+                    .put(NEW_COPY_ON_WRITE_ARRAY_SET, CopyOnWriteArraySet.class)
+                    .put(NEW_COPY_ON_WRITE_ARRAY_SET_WITH_ITERABLE, CopyOnWriteArraySet.class)
+                    .put(NEW_LINKED_HASH_SET, LinkedHashSet.class)
+                    .put(NEW_LINKED_HASH_SET_WITH_ITERABLE, LinkedHashSet.class)
+                    .put(NEW_LINKED_HASH_SET_WITH_EXPECTED_SIZE, LinkedHashSet.class)
+                    .put(NEW_TREE_SET, TreeSet.class)
+                    .put(NEW_TREE_SET_WITH_COMPARATOR, TreeSet.class)
+                    .put(NEW_TREE_SET_WITH_ITERABLE, TreeSet.class)
+                    .put(NEW_HASH_SET_WITH_ITERABLE, HashSet.class)
+                    .put(NEW_TREE_MAP_WITH_SORTED_MAP, TreeMap.class)
+                    .put(NEW_TREE_MAP_WITH_COMPARATOR, TreeMap.class)
+                    .put(NEW_LINKED_HASH_MAP_WITH_MAP, LinkedHashMap.class)
+                    .put(NEW_LINKED_HASH_MAP_WITH_EXPECTED_SIZE, LinkedHashMap.class)
+                    .put(NEW_ENUM_MAP, EnumMap.class)
+                    .put(NEW_ENUM_MAP_WITH_CLASS, EnumMap.class)
+                    .put(NEW_ENUM_MAP_WITH_MAP, EnumMap.class)
+                    .put(NEW_IDENTITY_HASH_MAP, IdentityHashMap.class)
+                    .build();
+
+    private static final Set<Matcher<ExpressionTree>> requiresCollectionArg = ImmutableSet.of(
+            NEW_ARRAY_LIST_WITH_ITERABLE,
+            NEW_LINKED_LIST_WITH_ITERABLE,
+            NEW_COPY_ON_WRITE_ARRAY_LIST_WITH_ITERABLE,
+            NEW_COPY_ON_WRITE_ARRAY_SET_WITH_ITERABLE,
+            NEW_LINKED_HASH_SET_WITH_ITERABLE,
+            NEW_TREE_SET_WITH_ITERABLE,
+            NEW_HASH_SET_WITH_ITERABLE);
+
+    @Override
+    public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
+
+        Class<?> collectionClass = findCollectionClassToUse(state, tree);
+        if (collectionClass == null) {
+            return Description.NO_MATCH;
+        }
+
+        SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
+        String collectionType = SuggestedFixes.qualifyType(state, fixBuilder, collectionClass.getName());
+        String typeArgs = tree.getTypeArguments()
+                .stream()
+                .map(state::getSourceForNode)
+                .collect(Collectors.joining(", "));
+        String arg = tree.getArguments().isEmpty() ? "" : state.getSourceForNode(tree.getArguments().get(0));
+        String replacement = "new " + collectionType + "<" + typeArgs + ">(" + arg + ")";
+        return buildDescription(tree)
+                .setMessage("The factory method call should be replaced with a constructor call. "
+                        + "See https://git.io/JeCT6 for more information.")
```

Comment:
curious why opaque shortened link?

```suggestion
                        + "See https://github.com/palantir/gradle-baseline/blob/develop/docs/best-practices/java-coding-guidelines/readme.md#avoid-generics-clutter-where-possible for more information.")
```